### DISCUSSION TITLE
Validate Eclipse product file using Maven/Tycho

### DISCUSSION THREAD

**Discussion by Prithvi2407** [eclipse-tycho/tycho#4028] [Q&A]
I'm working on an RCP application that uses a feature based product file and I would like to know if it possible to validate the product file with the same function as in the PDE UI. In the product file editor it is the “validate…” icon in the tool bar. I would like to have the Tycho build to fail if the product does not validate. (Fail early principle). So far I haven't found any solutions, I know there is org.eclipse.tycho.extras/target-platform-validation-plugin specifically for the target platform validation but is there anything similar to validate a product file and have the Tycho build fail if there any missing plugins/dependencies in the product? Here's an example of my pom.xml file -
```
             <plugin>
                <groupId>org.eclipse.tycho</groupId>
                <artifactId>tycho-p2-director-plugin</artifactId>
                <version>${tycho.version}</version>
                <executions>
                    <execution>
                        <id>materialize-products</id>
                        <goals>
                            <goal>materialize-products</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>archive-products</id>
                        <phase>pre-integration-test</phase>
                        <goals>
                            <goal>archive-products</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <products>
                        <product>
                            <id>com.gui.product</id>
                            <attachId>GUI</attachId>
                            <archiveFileName>com.gui.product-${project.version}-${buildqualifier}
                            </archiveFileName>
                        </product>
                    </products>
                </configuration>
            </plugin>
```
I tried adding ```<phase>verify</phase>``` to materialize-products goal as per some suggestions I found but that didn't help either.
Is this something that can be achieved?

**Comment by laeubi** [comment]
If a product is "valid" can't be determined, even if there are some missing dependencies it might be intentional.
Beside that, you usually get some good validation of completeness if you use a feature base product as features are validated during build and for products it requires all features to be there.

**Reply by Prithvi2407** [reply]
My product is feature based and some features may have some missing dependencies which Eclipse's "validate..." button reports correctly but Tycho adds the missing dependencies silently during the build so is there a way to fail the Tycho build in such scenarios?

**Reply by merks** [reply]
In the IDE the validation is based on what's available in the target platform in the running IDE.   It's not clear how you specified the target platform for your Tycho build.  If it's finding all the necessary transitive dependencies then there are none missing for your target platform during the Tycho build...

**Reply by laeubi** [reply]
As mentioned, if anything would be missing then a feature based product would fails to build. Also recent PDE have the feature to include what is required automatically, so maybe you just need to update your Eclipse IDE?

**Reply by Prithvi2407** [reply]
Sorry if there was a misunderstanding. Actually what I'm trying to achieve is to have more control over Tycho and make the build fail if the features are incomplete (due to missing dependencies) so that the features are inclusive. 
Basically, PDE provides an option to validate the .product file and performs a validation check on the .product file to ensure that all features and plugins specified in the product configuration are correctly set up and that there are no missing dependencies through "Validate.." button in .product file editor so I would like to achieve the same with Tycho build where Tycho should report such an issue and the build should fail.

**Reply by laeubi** [reply]
Actually these "complete" features are bad, Tycho never (or long time...) not required features to be "complete" because requirements are simply pulled in automatically, additionally you can't upgrade such "complete" features easily. Now PDE has a new option to behave like Tycho: 
![grafik](https://github.com/eclipse-tycho/tycho/assets/1331477/bc35bd07-33a7-4e32-aed2-30165a55da03)

**Comment by merks** [comment]
A feature being complete is a separate aspect from a product being complete.  A product might include two features, both of which are incomplete but the union of the two is complete.     E.g, two features F1 and F2 might transitively both require bundles X and Y but one includes only X and other only Y.  The product including F1 and F2 would be complete  but neither feature is complete.

It's just not entirely clear what exactly you are trying to achieve and more importantly why you are trying to achieve.  Maybe you perceive feature completeness to be a solution to some problem that is likely solved in some other way or is not actually even a problem.   Most of the Eclipse SDK's features are not complete in the sense you are describing without that actually being a problem....

**Reply by laeubi** [reply]
One should especially keep in mind that:

1. even if a features does not list "everything" and is installed, P2 will install what ever is required
2. even if a product does include features that are "incomplete" it will package everything what is required

So features and products must only list those dependencies that are the "root" or not referenced/required anywhere and that's even much easier to maintain, so as @merks mentined, what is the problem that needs to be solved (beside that PDE complains if you disable that option).

**Reply by Prithvi2407** [reply]
We want to have control over what P2 installs for the product. 
Basically, we want to keep one of our products as minimal as possible because of installation/memory footprint, startup time (number of bundles to scan), etc. 
This is why we use the IDE Validation to check whether the product is complete even without installing implicitly required bundles.
Otherwise we'd need to implement our own check to, e.g., analyze the generated bundles.info file that it did not add any unwanted bundle violating our "minimal product" requirements. Having tycho behave like the strict IDE check would solve this issue for us.

**Reply by laeubi** [reply]
> We want to have control over what P2 installs for the product.

As far as I know there is no such way, P2 basically installs everything that is required even if you not mention it in a feature.

> Basically, we want to keep one of our products as minimal as possible because of installation/memory footprint, startup time (number of bundles to scan), etc.

Then you should look at the resulting product not at the features, especially adding things to a feature that is not required (because it is already a dependency) tends to make this task even harder as its easy to forget that a dependency is no longer needed.

> This is why we use the IDE Validation to check whether the product is complete even without installing implicitly required bundles.

As said this is an (old) feature thats not required anymore these days (from my point of view), also in the IDE nothing is "installed", it is just selected, so what is really installed (e.g. launcher artifacts, JVM, ...) you can only see in a final build product.

> Otherwise we'd need to implement our own check to, e.g., analyze the generated bundles.info file that it did not add any unwanted bundle violating our "minimal product" requirements.

Tycho always builds a minimal product unless you explicitly adds things to the feature, so in turn you should better try to *remove* things from features instead of *adding* things to features.

> Having tycho behave like the strict IDE check would solve this issue for us.

The IDE check is not "strickt" (see explanation above) it just has previously required you to do things manually that (in the IDE) that Tycho/P2 already performs automatically....

**Reply by merks** [reply]
How exactly do you specify the target platform for your build?   I ask because if you list all units explicitly in the *.target and use slicer mode, not planner mode, you would notice if some new requirement was not specified explicitly...

**Reply by xpomul** [reply]
Hi Ed @merks,

the thing is, there is a company-wide target platform (very large) for all kinds of products. 
We have a very specific, small, product (basically a CLI client that communicates with a server, but still based on the general TP) that contains just a few bundles with a minimal set of dependencies (Equinox, org.eclipse.core.runtime, ...) + some bundles shared between the minimal product and other (bigger) products. 
We have also defined our own minimal Eclipse Platform feature containing only that minimal set of Platform bundles, so that the PDE Product Validation succeeds without complaints and without adding implicit dependencies.

If now someone added a dependency in the common bundle to, e.g., org.eclipse.core.resources, then tycho-p2 would accept that and we would end up with a built product that contains several additional bundles (higher footprint, the workspace which we don't need is initialized, etc.).

Until now we have (mis-?)used the Eclipse PDE Product Validation to detect that our minimal feature is not sufficient anymore (the validation would report the org.eclipse.core.resources bundle as required but missing, and we know that someone added an unwanted dependency. But as clicking the "Validate"-button in the IDE is a manual task that can be easily forgotten, we thought it might be nice to have the CI build fail in case that validation fails. So we were checking for a way to realize that...

But as this seems to be deprecated/not supported, we will write a custom build step now that checks the content of the bundles.info in the built product instead to address our issue.

Anyways, thanks for the hints and support :-)

**Comment by laeubi** [comment]
So just to clarify, assume you have a product and it contains **one feature** and that includes **one bundle** then Tycho will build a minimal product that only includes what the bundle requires, so no manual maintaining is required.

Furthermore assume there is an updatesite and it contains **one feature** and that includes **one bundle** and you install it into an exiting IDE/Product, P2 will install this bundle and includes what the bundle requires, so no manual maintaining is required. 

So you can only control what is minimal installed not what is installed in total...

The same thing can be archived when building an updatesite with varying degrees of "size" e.g. using references to existing sites.
